src/app/centroidController.js

```diff
@@ -0,0 +1,261 @@
+
+              this.centroidLayer.setVisibility(false);
+
+              this.centroidLayer.on('load', d.resolve);
+
+              deferreds.push(d);
+
+              [li.poly, li.line, li.point].forEach(function (layerIndex, i) {
+                  var layer = new FeatureLayer(config.urls.mapService + '/' + layerIndex, {
+                      id: typesLookup[i],
+                      mode: FeatureLayer.MODE_SELECTION
+                  });
+
+                  layer.setVisibility(false);
+
+                  var deferred = new Deferred();
+
+                  layer.on('load', deferred.resolve);
+
+                  deferreds.push(deferred);
+
+                  this.explodedLayer[typesLookup[i]] = layer;
+              }, this);
+
+              this.layersLoaded = true;
+
+              return all(deferreds);
+          }
+
+          return true;
+      },
+      updateLayerVisibilityFor: function (extent) {
+          // summary:
+          //      mimics the extent change event to trigger the decision tree of whether to show or hide centroids
+          // extent: {5:type or return: type}
+          console.log('app.mapControls.centroidController::updateLayerVisibilityFor', arguments);
+
+          if (!extent) {
+              return;
+          }
+
+          this.onExtentChanged(extent);
+      },
+      showFeaturesFor: function (where) {
+          // summary:
+          //      selects features for the where clause
+          // where: {5:type or return: type}
+          console.log('app.centroidController::showFeaturesFor', arguments);
+
+          var q = new Query();
+          q.where = where || this.where;
+
+          this.where = where;
+
+          var deferreds = [];
+
+          if (this.centroidsVisible) {
+              deferreds.push(this.centroidLayer.selectFeatures(q));
+          } else {
+              Object.keys(this.explodedLayer).forEach(function (key) {
+                  var layer = this.explodedLayer[key];
+                  deferreds.push(layer.selectFeatures(q));
+              }, this);
+          }
+
+          if (this.where === '1 = 1') {
+              return null;
+          }
+
+          return all(deferreds).then(
+              function (graphics) {
+                  if (graphics === null) {
+                      // state of utah extent
+                      return null;
+                  } else {
+                      return graphicsUtils.unionGraphicsIntoExtent(graphics);
+                  }
+              }
+          );
+      },
+      _showPopupForProject: function (show, evt) {
+          // summary:
+          //      shows the dialog popup
+          // evt: https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent
+          console.log('app.centroidController::_showPopupForProject', arguments);
+
+          if (!show) {
+              popup.close();
+              return;
+          }
+
+          this.dialog.set('content', lang.replace('<div class="project-details">' +
```

Comment:
This would be easier to read and maintain if it's in an external `.html` file and brought in via `dojo/text!`.